adapters/telaria/telaria.go

```diff
@@ -0,0 +1,331 @@
+				Message: "Telaria doesn't support banner",
+			}
+		}
+
+		hasVideoObject = hasVideoObject || imp.Video != nil
+	}
+
+	if !hasVideoObject {
+		return &errortypes.BadInput{
+			Message: "No Video object present in Imp object",
+		}
+	}
+
+	return nil
+}
+
+// Fetches the populated header object
+func GetHeaders(request *openrtb.BidRequest) *http.Header {
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	headers.Add("X-Openrtb-Version", "2.5")
+	headers.Add("Accept-Encoding", "gzip")
+
+	if request.Device != nil {
+		if len(request.Device.UA) > 0 {
+			headers.Add("User-Agent", request.Device.UA)
+		}
+
+		if len(request.Device.IP) > 0 {
+			headers.Add("X-Forwarded-For", request.Device.IP)
+		}
+
+		if len(request.Device.Language) > 0 {
+			headers.Add("Accept-Language", request.Device.Language)
+		}
+
+		if request.Device.DNT != nil {
+			headers.Add("Dnt", strconv.Itoa(int(*request.Device.DNT)))
+		}
+	}
+
+	return &headers
+}
+
+// Checks the imp[i].ext object and returns a imp.ext object as per ExtImpTelaria format
+func (a *TelariaAdapter) FetchTelariaExtImpParams(imp *openrtb.Imp) (*openrtb_ext.ExtImpTelaria, error) {
+	var bidderExt adapters.ExtImpBidder
+	err := json.Unmarshal(imp.Ext, &bidderExt)
+
+	if err != nil {
+		err = &errortypes.BadInput{
+			Message: "ext.bidder not provided",
+		}
+
+		return nil, err
+	}
+
+	var telariaExt openrtb_ext.ExtImpTelaria
+	_ = json.Unmarshal(bidderExt.Bidder, &telariaExt)
+
+	return &telariaExt, nil
+}
+
+// Fetch the id from the appropriate publisher object
+func (a *TelariaAdapter) FetchPublisherId(request *openrtb.BidRequest) string {
+
+	if request.Site != nil {
+		if request.Site.Publisher != nil {
+			if request.Site.Publisher.ID != "" {
+				return request.Site.Publisher.ID
+			}
+		}
+	}
+
+	if request.App != nil {
+		if request.App.Publisher != nil {
+			if request.App.Publisher.ID != "" {
+				return request.App.Publisher.ID
+			}
+		}
+	}
+
+	return ""
+}
+
+// This method changes <site/app>.publisher.id to request.ext.telaria.seatCode
+// And moves the publisher.id to request.ext.originalPublisherId
+func (a *TelariaAdapter) PopulateRequestExtAndPubId(request *openrtb.BidRequest) error {
+	var requestExtIncoming ReqExtIn
+	if err := json.Unmarshal(request.Ext, &requestExtIncoming); err != nil {
+		return err
+	}
+
+	if requestExtIncoming.Telaria.SeatCode == "" {
+		return &errortypes.BadInput{Message: "Seat Code is required"}
+	}
+
+	request.Ext, _ = json.Marshal(&BidExtOut{requestExtIncoming.ExtRequest, a.FetchPublisherId(request)})
+
+	publisherObject := &openrtb.Publisher{ID: requestExtIncoming.Telaria.SeatCode}
+
+	if request.Site != nil {
+		if request.Site.Publisher != nil {
+			request.Site.Publisher.ID = requestExtIncoming.Telaria.SeatCode
+		} else {
+			request.Site.Publisher = publisherObject
+		}
+	}
+
+	if request.App != nil {
+		if request.App.Publisher != nil {
```

Comment:
Both `request.Site.Publisher` and `request.App.Publisher` get checked for `nil` inside `FetchPublisherId(request *openrtb.BidRequest)` and checked again in this function. Is there a way to avoid the extra check in lines 178 to 192? I guess it translates to something like:
```
141   func (a *TelariaAdapter) FetchPublisherId(request *openrtb.BidRequest) (string) {
142 - 
    +     pubID := requestExtIncoming.Telaria.SeatCode
143       if request.Site != nil {
144           if request.Site.Publisher != nil {
145 -             if request.Site.Publisher.ID != "" {
146 -                 return request.Site.Publisher.ID
147 -             }
    +             request.Site.Publisher.ID = pubID
    +         } else {
    +             pubID = request.Site.Publisher.ID
148           }
149       }
150  
151       if request.App != nil {
152           if request.App.Publisher != nil {
153 -             if request.App.Publisher.ID != "" {
154 -                 return request.App.Publisher.ID
155 -             }
    +             request.App.Publisher.ID = pubID
    +         } else {
    +             pubID = request.App.Publisher.ID
156           }
157       }
158  
159 -     return ""
    +     return pubID
160   }
161  
162   // This method changes <site/app>.publisher.id to request.ext.telaria.seatCode
163   // And moves the publisher.id to request.ext.originalPublisherId
164   func (a *TelariaAdapter) PopulateRequestExtAndPubId(request *openrtb.BidRequest) error {
165       var requestExtIncoming ReqExtIn
166       if err := json.Unmarshal(request.Ext, &requestExtIncoming); err != nil {
167           return err
168       }
169  
170       if requestExtIncoming.Telaria.SeatCode == "" {
171           return &errortypes.BadInput{Message: "Seat Code is required"}
172       }
173  
    +     originalPublisherID := a.FetchPublisherId(request)
174 -     request.Ext, _ = json.Marshal(&BidExtOut{requestExtIncoming.ExtRequest, a.FetchPublisherId(request)})
    +     request.Ext, _ = json.Marshal(&BidExtOut{requestExtIncoming.ExtRequest, originalPublisherID})
175  
176       publisherObject := &openrtb.Publisher{ID: originalPublisherID}
177 -
178 -     if request.Site != nil {
179 -         if request.Site.Publisher != nil {
180 -             request.Site.Publisher.ID = requestExtIncoming.Telaria.SeatCode
181 -         } else {
182 -             request.Site.Publisher = publisherObject
183 -         }
184 -     }
185 -
186 -     if request.App != nil {
187 -         if request.App.Publisher != nil {
188 -             request.App.Publisher.ID = requestExtIncoming.Telaria.SeatCode
189 -         } else {
190 -             request.App.Publisher = publisherObject
191 -         }
192 -     }
193  
194       return nil
195   }
```
This is just a suggestion and the logic needs to be double checked.

Reply:
Right, I've combined the nil checks and the seat code assignments.